presets/ragengine/vector_store/base.py

```diff
@@ -208,6 +208,33 @@ async def add_document_to_index(self, index_name: str, document: Document, doc_i
     def list_indexes(self) -> List[str]:
         return list(self.index_map.keys())
 
+    async def delete_document(self, index_name: str, doc_id: str):
+        """Common logic for deleting a document."""
+        if index_name not in self.index_map:
+            raise ValueError(f"No such index: '{index_name}' exists.")
+        if self.use_rwlock:
+            async with self.rwlock.writer_lock:
+                self.index_map[index_name].delete_ref_doc(doc_id, delete_from_docstore=True)
+                self.index_store.add_index_struct(self.index_map[index_name].index_struct)
+        else:
+            self.index_map[index_name].delete_ref_doc(doc_id, delete_from_docstore=True)
+            self.index_store.add_index_struct(self.index_map[index_name].index_struct)
+    
+    async def update_document(self, index_name: str, doc_id: str, document: Document):
+        """Common logic for updating a document."""
+        if index_name not in self.index_map:
+            raise ValueError(f"No such index: '{index_name}' exists.")
+        
+        llama_doc = LlamaDocument(id_=doc_id, text=document.text, metadata=document.metadata, excluded_llm_metadata_keys=[key for key in document.metadata.keys()])
+
+        if self.use_rwlock:
+            async with self.rwlock.writer_lock:
+                await self.index_map[index_name].update_ref_doc(llama_doc)
+                self.index_store.add_index_struct(self.index_map[index_name].index_struct)
+        else:
+            await self.index_map[index_name].update_ref_doc(llama_doc)
+            self.index_store.add_index_struct(self.index_map[index_name].index_struct)
```

Comment:
Awesome. Same comments as above - "update is equivalent to deleting the document and then inserting it again" (https://docs.llamaindex.ai/en/stable/api_reference/indices/) so if delete for example isn't implemented neither with update.
    
mini example
        try:
            self.index_map[index_name].update_ref_doc(llama_doc)
        except NotImplementedError as e:
            print("Update is equivalent to deleting the document and then inserting it again.")
            print(f"Update not yet fully implemented for index. Skipping document {document.doc_id}. Error: {e}")
        except Exception as e:
            print(f"Unable to Update Document in the VectorStoreIndex. Skipping. Error: {e}")